Automate `elastic_constants` by AD-DFPT as a postprocessing

niklasschmitz opened the pull request:
Following up on the ideas at #1162.

This PR aims to automate the clamped-ion elastic constants as a convenient postprocessing function `elastic_constants(scfres)` for 1) the generic case of not using symmetry (most expensive, 6xDFPT on a fully unfolded kgrid...), 2.1) the most common case of cubic crystals (least expensive, 1xDFPT on a partially unfolded kgrid). Optimizations for other the point groups could be implemented later following a similar pattern.

progress:
- [x] Implement generic case
- [x] Implement cubic case
- [x] use SCF results from scfres and partially unfold (#1308; Update: just warm-starting with the correct density and tightened diagonalization tolerance)
- [x] more testing
- [x] mention in elastic constants docs example

niklasschmitz committed:
Start elastic_constants draft

mfherbst commented on src/postprocess/elastic.jl:
```diff
@@ -0,0 +1,59 @@
+end
+
+"""Computes the *clamped-ion* elastic constants (without ionic relaxation)"""
+function elastic_constants(scfres::NamedTuple; 
```
How about `elastic_tensor` because that's what it is. The constants are the individual entries.

mfherbst commented on src/postprocess/elastic.jl:
```diff
@@ -0,0 +1,59 @@
+    # TODO restart SCF from previous
+    lattice = DFTK.voigt_strain_to_full(voigt_strain) * model0.lattice
+    model = Model(model0; lattice, symmetries)
+    basis = PlaneWaveBasis(model; Ecut, kgrid)
```
This is dangerous as basis of the SCF may be very different (e.g. fft_grid may be set, symmetries may be customised, architecture etc.). We have to find a better mechanics here. I would be in favour of having a special PlaneWaveBasis constructor that essentially just changes the model under the hood and otherwise tries to keep as many of the settings of the "old basis". Would also be useful for https://github.com/JuliaMolSim/DFTK.jl/blob/master/src/postprocess/stresses.jl#L22.

niklasschmitz committed:
Rename elastic_constants -> elastic_tensor

niklasschmitz committed:
Merge branch 'master' into elastic-postprocess

niklasschmitz committed:
Add elastic test

niklasschmitz committed:
Merge branch 'master' into elastic-postprocess

niklasschmitz committed:
Review

niklasschmitz committed:
Simplify

niklasschmitz committed:
Add PlaneWaveBasis model update constructor

niklasschmitz committed:
Merge branch 'master' into elastic-postprocess

niklasschmitz committed:
Address review

niklasschmitz committed:
Line length

niklasschmitz committed:
Merge branch 'master' into elastic-postprocess

niklasschmitz commented on src/postprocess/elastic.jl:
```diff
@@ -0,0 +1,67 @@
+full Jacobian is computed.
+"""
+function elastic_tensor(scfres::NamedTuple;
+                        magnetic_moments=[],
```
opened #1307

mfherbst replied:
Ok, but this needs a todo here as well.

niklasschmitz commented:
> Something that bothers me as well is that it's basically impossible to send any kind of parameters to the DFPT solver in your implementation. There is the `ResponseOptions` for this purpose. Maybe just taking one as a keyword argument from the elastic constant wrapper function would be simple solution ?

Agreed. I now added the more general solution to pass arbitrary `kwargs_scf...` to `elastic_tensor`.

> The displacement `0.01` should depend on the `SYMMETRY_TOLERANCE`

 Makese sense. I added tol_symmetry as keyword argument and displace by `100 * tol_symmetry` (the `1e-5` default `SYMMETRY_TOLERANCE` gets `1e-3` strain).

niklasschmitz committed:
Mention `elastic_tensor` in docs

mfherbst commented on examples/elastic_constants.jl:
```diff
@@ -152,6 +152,21 @@ println("C44 (FD): ", uconvert(u"GPa", dstress_fd[4] * u"hartree" / u"bohr"^3))
+println("C12: ", uconvert(u"GPa", C[1, 2] * u"hartree" / u"bohr"^3))
+println("C44: ", uconvert(u"GPa", C[4, 4] * u"hartree" / u"bohr"^3))
+
 # ## Moving to meta-GGA functionals
```
Use your function here instead of the manual shenanigans and move up as well.

niklasschmitz replied:
done

mfherbst commented on examples/elastic_constants.jl:
```diff
@@ -152,6 +152,21 @@ println("C44 (FD): ", uconvert(u"GPa", dstress_fd[4] * u"hartree" / u"bohr"^3))
 # Note the discrepancy in c44, which is due to us not yet including
 # ionic relaxation in this example.
 
+# ## Using the `elastic_tensor` postprocessing function
```
If I was a practitioner, this is what I care about not any of the things written above (which is "specialist's knowledge".) So this should come first. In fact, there may even be a point to move out the detailed instructions (same as for the polarisability) and make this a separate thing ... but this can come later.

niklasschmitz replied:
done

mfherbst commented on src/postprocess/elastic.jl:
```diff
@@ -0,0 +1,81 @@
+"""
+    elastic_tensor(scfres; magnetic_moments=[],
+                   tol_symmetry=SYMMETRY_TOLERANCE,
+                   kwargs_scf...)
```
I am not so sure. This is a very surprising interface. Imagine you converge the scf to 1e-12. You expect the elastic tensor calculation to respect this accuracy, but it won't. We should build a function to automatically extract the relevant SCF kwargs from the `scfres` for convergence control (`is_converged`, `mixing`, `damping`, ...) and use that here.

Regarding the kwarg, since this is a response step that comes after the SCF step, the only reasonable kwarg in my opinion is `ResponseOptions` (which should become the standard way to pass kwargs to the implicit `solve_ΩplusK_split` executed when an SCF is differentiated).

We can also settle on another pattern, but I think now is the moment we should work towards one. Because if we don't it quickly gets super messy. If all functions take just kwargs and these get decomposed along the call stack, we will loose the overview which kwargs can be set from where.

niklasschmitz replied:
Agreed, I put `ResponseOptions` first now as kwarg, the other options are just workarounds.

I now inlined a scavenging of the scfres for SCF-related options, a temporary fix with a TODO. A cleaner option could be to bundle SCF options together in an scfres struct.

Even the `tol_symmetry` kwarg for `elastic_tensor` one might discuss whether it can be removed - to me it would make sense to store tol_symmetry as part of the model next to the symmetry group information (i.e. the symmetry result can be bundled with its uncertainty) then we don't have to have it as kwarg here and can just read it out of `scfres.basis.model`.

mfherbst replied:
> store tol_symmetry as part of the model next to the symmetry group information

Yes, can you add this to the other issue ?

niklasschmitz replied:
done

niklasschmitz committed:
Update docs

niklasschmitz committed:
Address review + warm-start density

mfherbst commented on src/postprocess/elastic.jl:
```diff
@@ -2,39 +2,58 @@ import DifferentiationInterface as DI
     model = Model(model0; lattice, symmetries)
     basis = PlaneWaveBasis(basis0; model)
-    scfres = self_consistent_field(basis; kwargs_scf...)
+    scfres = self_consistent_field(basis; ρ=something(ρ, guess_density(basis)),
```
`guess_density` needs the magnetic moments. I think it's fine to have `ρ` be a mandatory kwarg in this internal function, what do you think ?

niklasschmitz replied:
Yes I think mandatory \rho is fine here, changed it

niklasschmitz committed:
For _stress_from_strain, require ρ (simpler)

niklasschmitz committed:
Merge branch 'master' into elastic-postprocess

niklasschmitz committed:
Fix scfres_to_dict! with new fields

mfherbst commented on src/input_output.jl:
```diff
@@ -369,6 +370,11 @@ function scfres_to_dict!(dict, scfres::NamedTuple; save_ψ=true, save_ρ=true)
         if hasproperty(scfres, :mixing)
             dict["mixing"] = string(scfres.mixing)
         end
+        for sym in (:is_converged, :nbandsalg, :fermialg, :diagtolalg, :solver, :eigensolver)
+            if hasproperty(scfres, sym)
+                dict[string(sym)] = string(getproperty(scfres, sym))
+            end
+        end
```
This is code duplication and I'm not sure if the strings you get out of `solver` and `eigensolver` are very useful. E.g. eigensolver could just be a large bunch of numbers (which could blow up the size of the json). Have you checked that ?

niklasschmitz replied:
It seems to work fine for the function arguments, it just prints the name. On `examples/silicon.jl` it looks like this:
```
julia> dict = Dict()
Dict{Any, Any}()

julia> for sym in (:is_converged, :nbandsalg, :fermialg, :diagtolalg, :solver, :eigensolver)
     if hasproperty(scfres, sym)
         dict[string(sym)] = string(getproperty(scfres, sym))
     end
 end

julia> dict
Dict{Any, Any} with 6 entries:
  "nbandsalg"    => "AdaptiveBands(4, 7, 1.0e-6, 0.01)"
  "eigensolver"  => "lobpcg_hyper"
  "is_converged" => "ScfConvergenceDensity(1.0e-8)"
  "diagtolalg"   => "AdaptiveDiagtol(0.2, nothing, 0.005, 0.03)"
  "solver"       => "anderson"
  "fermialg"     => "FermiTwoStage()"
```

mfherbst replied:
Perfect, great, thanks for checking !
